### ISSUE TITLE
Nullable type

### ISSUE THREAD

**Issue by Jaidsuka** [commandlineparser/commandline#305]
IsEmpty method doesn't work right in case when we use Nullable type as option. In this case when we pass 0 to parameter it won't add this option to command line back because think that default value for it will be 0 too. Nullable<'int'>.GetType() == typeof(int). IsEmpty method should be changed for covering that problem.

**Comment by moh-hassan** [comment]
>when we pass 0 to parameter it won't add this option to command line back because think that default value for it will be 0

if you defined an option as nullable<int> like:

           class Option
          {
            [Option('v',  HelpText = "Verbose level. Range: from 0 to 2.")]
            public int? VerboseLevel { get; set; }
         }

 When 0  is passed  in the command line: `  myapp -v 0 ` , the `Option.VerboseLevel`   will be 0

For Nullable<int> , 0 isn't the default value.

BTW, the default value of a nullable type is an instance for which the HasValue property is false and the Value property is undefined.
So, to check that VerboseLevel  is assigned a value in the commandline, you can check

               VerboseLevel .HasValue // true if assihned a value in the commandline

**Comment by tydunkel** [comment]
@moh-hassan I am running into the same issue. @Jaidsuka is pointing to the `IsEmpty` method in [UnParserExtensions](https://github.com/commandlineparser/commandline/blob/d16d24c4eef664bc4d297b193d9d79478f9ce96a/src/CommandLine/UnParserExtensions.cs#L245).

Using your above example, having an option defined as:
```
      class Option
      {
        [Option('v',  HelpText = "Verbose level. Range: from 0 to 2.")]
        public int? VerboseLevel { get; set; }
     }
```
Calling `myapp -v 0` will set the verbose level properly, but then calling `parser.FormatCommandLine(options)` with those options (VerboseLevel=0) will return an empty string. This is because the `IsEmpty` method will return `true` for a Nullable<int> when the value is 0.

This means that anyone relying on `FormatCommandLine` to rebuild the original commandline will not be able to do so if their options contain nullable types.

**Comment by moh-hassan** [comment]
PR is welcome to fix this issue.

**Comment by nemars** [comment]
How about something like this? This is also helping with the issue 379:
https://github.com/commandlineparser/commandline/issues/379
https://github.com/commandlineparser/commandline/pull/452

        private static bool IsEmpty(object value, Type type)
        {
            if (value == null) return true;
            if (ReflectionHelper.IsFSharpOptionType(type) && !FSharpOptionHelper.IsSome(value)) return true;
            if (type.IsValueType && value.Equals(type.GetDefaultValue())) return true;
            if (value is string && ((string)value).Length == 0) return true;
            if (value is IEnumerable && !((IEnumerable)value).GetEnumerator().MoveNext()) return true;
            return false;
        }

NOTICE: I removed "#if !SKIP_FSHARP" and "#endif" because this comment was being formatted as text.

**Comment by tydunkel** [comment]
@nemars I don't think that will work. `type` will be a `ValueType` with a non-null default. This is the root of the issue. We need to determine if the type in the `OptionsSpecification` is actually a `Nullable` value type.

**Comment by nemars** [comment]
That's why I put both value and type as parameters. This is called only in one place in the code and the type can easily be acquired before that call.

pi => new { 
    Specification = Specification.FromProperty(pi),
    Value = pi.GetValue(options, null).NormalizeValue(),
    PropertyValue = pi.GetValue(options, null,
    **pi.PropertyType)**
}

**Comment by moh-hassan** [comment]
[online demo](https://dotnetfiddle.net/8gPgBK) to show the fix in v2.7+
